### PR TITLE
Add custom programs to ProgramCache

### PR DESCRIPTION
**Pull request by bolino**
Store and retrieve custom user-defined programs in ProgramCache.

This feature was in v3 but not present in v4.

### CODE REVIEW

**Review by Mee-gu — APPROVED** [review]

File: cocos/renderer/backend/ProgramCache.cpp [hunk 1]
```diff
@@ -303,6 +305,29 @@ void ProgramCache::removeAllPrograms()
...
+    for (auto& program : _cachedCustomPrograms)
+    {
+        program.second->release();
+    }
+    _cachedCustomPrograms.clear();
+}
+
+void ProgramCache::addCustomProgram(const std::string &key, backend::Program *program)
+{
+    _cachedCustomPrograms.emplace(key, program);
```
Comment (Mee-gu) [inline cocos/renderer/backend/ProgramCache.cpp:318]: need release old program if key is already exist in `_cachedCustomPrograms`.

Comment (bolino) [reply cocos/renderer/backend/ProgramCache.cpp:318]: I have fixed this in https://github.com/ironhidegames/cocos2d-x/commit/69f393333a72d1956f589199209069637a157067
Should I make a new PR?

Comment (minggo) [reply cocos/renderer/backend/ProgramCache.cpp:318]: Yes, please.

Comment (Mee-gu) [reply cocos/renderer/backend/ProgramCache.cpp:318]: Yes, it would be great.

Comment (bolino) [reply cocos/renderer/backend/ProgramCache.cpp:318]: I have made a new PR: https://github.com/cocos2d/cocos2d-x/pull/20550

**Pull request merged**